hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,260 @@
+
+    def cancel(self):
+        if os.getpid() == self._origin_pid:
+            self.active_futures.pop(self._uid, None)
+            return super().cancel()
+        elif self._state in [base.RUNNING, base.FINISHED]:
+            return False
+        else:
+            self._state_cache[self._state] = base.CANCELLED
+            self._send_update(UpdateType.CANCEL)
 
     def set_running_or_notify_cancel(self):
-        self._sync_updates()
         if self._state == base.PENDING:
             self._state = base.RUNNING
-            return self._send_updates()
+            return True
         elif self._state == base.CANCELLED:
             return False
         else:
-            raise FutureStateError(f"Can't set_running_or_notify_cancel to a future that is in {self._state} ({self})")
-
-    def cancel(self):
-        self._sync_updates()
-        if self._state in self.TERMINAL_STATES:
-            return False
-        self._state, self._exception = base.CANCELLED, base.CancelledError()
-        return self._send_updates()
+            raise InvalidStateError(f"Can't set_running_or_notify_cancel when future is in {self._state} ({self._uid})")
 
     def result(self, timeout: Optional[float] = None) -> ResultType:
-        self._await_terminal_state(timeout)
-        if self._exception is not None:
+        if self._state not in TERMINAL_STATES:
+            assert os.getpid() == self._origin_pid, "only the process that created MPFuture can await result."
+            return super().result(timeout)
+        elif self._state == base.CANCELLED:
+            raise base.CancelledError()
+        elif self._exception:
             raise self._exception
-        return self._result
+        else:
+            return self._result
 
-    def exception(self, timeout=None) -> BaseException:
-        self._await_terminal_state(timeout)
-        if self._state == base.CANCELLED:
+    def exception(self, timeout: Optional[float] = None) -> BaseException:
+        if self._state not in TERMINAL_STATES:
+            assert os.getpid() == self._origin_pid, "only the process that created MPFuture can await exception."
+            return super().exception(timeout)
+        elif self._state == base.CANCELLED:
             raise base.CancelledError()
         return self._exception
 
     def done(self) -> bool:
-        self._sync_updates()
-        return self._state in self.TERMINAL_STATES
+        return self._state in TERMINAL_STATES
 
     def running(self):
-        self._sync_updates()
         return self._state == base.RUNNING
 
     def cancelled(self):
-        self._sync_updates()
         return self._state == base.CANCELLED
 
-    def add_done_callback(self, callback):
-        raise NotImplementedError(f"MPFuture doesn't support callbacks.")
-
-    def remove_done_callback(self, callback):
-        raise NotImplementedError(f"MPFuture doesn't support callbacks.")
+    def add_done_callback(self, callback: Callable):
```

Comment:
I'd suggest to make the expected signature clear:

```suggestion
    def add_done_callback(self, callback: Callable[['MPFuture'], None]):
```

Quotes around `MPFuture` are for Python 3.7.